File: app/dispatcher.py
"""
The core of the service: takes a DispatchRequest, fans out to all
destinations concurrently, retries failures with exponential backoff, and
records every attempt.
 
Concurrency model: asyncio.gather fires all destination deliveries at
once. Each destination gets its own retry loop running concurrently with
every other destination's retry loop. A failure at one destination (or its
retries) doesn't delay delivery to others.
"""

from __future__ import annotations
import asyncio
import datetime
import json
import logging

import httpx

from app.formatters import format_payload
from app.models import AttemptStatus, DeliveryAttempt, Destination, DestinationResult, DispatchRequest, DispatchResponse, EventRecord
from app.signing import sign
from app.store import EventStore

logger = logging.getLogger(__name__)

def _backoff(attempt: int, base: float, maximum: float) -> float:
    """Exponential backoff"""
    return min(base * (2 ** attempt), maximum)

async def _deliver_once(client: httpx.AsyncClient, url: str, body: dict, secret: str | None, timeout: float) -> tuple[int | None, str | None]:
    """Attempts a single HTTP POST. Returns (http_status, error_message)."""

    raw = json.dumps(body, default=str).encode()
    headers = {"Content-Type": "application/json"}
    if secret:
        headers["X-Webhook-Signature"] = sign(raw, secret)

    try:
        resp = await client.post(url, content=raw, headers=headers, timeout=timeout)
        if resp.is_success:
            return resp.status_code, None
        return resp.status_code, f"Non-2xx response: {resp.status_code}"
    except httpx.TimeoutException:
        return None, "Timed out"
    except httpx.HTTPError as e:
        return None, str(e)
    
async def _deliver_with_retries(
    client: httpx.AsyncClient,
    destination: Destination,
    body: dict,
    secret: str | None,
    max_attempts: int,
    backoff_base: float,
    backoff_max: float,
    delivery_timeout: float,
) -> DestinationResult:
    """Retries max_attempts times with exponential backoff"""
    url = str(destination.url)
    effective_secret = destination.secret or secret
    attempts: list[DeliveryAttempt] = []

    for attempt_num in range(1, max_attempts + 1):
        http_status, error = _deliver_once(client, url, body, effective_secret, delivery_timeout)

        status = AttemptStatus.success if error is None else AttemptStatus.failure
        attempts.append(DeliveryAttempt(
            attempt_number=attempt_num,
            timestamp=datetime.now(datetime.timezone.utc),
            status=status,
            http_status=http_status,
            error=error,
        ))

        if error is None:
            logger.info("Delivered to %s on attempt %d", url, attempt_num)
            break

        logger.warning("Delivery attempt %d to %s failed: %s", attempt_num, url, error)

        if attempt_num < max_attempts:
            delay = _backoff(attempt_num - 1, backoff_base, backoff_max)
            logger.info("Retrying %s in %.1f seconds", url, delay)
            await asyncio.sleep(delay)
    
    final_status = attempts[-1].status
    return DestinationResult(
        destination_type=destination.type,
        url=url,
        status=final_status,
        attempts=attempts,
    )

async def dispatch(
    request: DispatchRequest,
    event_id: str,
    client: httpx.AsyncClient,
    store: EventStore,
    max_attempts: int,
    backoff_base: float,
    backoff_max: float,
    delivery_timeout: float,
) -> DispatchResponse:
    """
    Fan-out: all destinations are delivered concurrently via asyncio.gather.
    Each gets its own formatted body.
    """

    async def deliver_to(destination: Destination) -> DestinationResult:
        body = format_payload(destination.type.value, request.payload, request.event_type)
        return await _deliver_with_retries(
            client=client,
            destination=destination,
            body=body,
            secret=request.secret,
            max_attempts=max_attempts,
            backoff_base=backoff_base,
            backoff_max=backoff_max,
            delivery_timeout=delivery_timeout,
        )
    
    results = await asyncio.gather(*[deliver_to(d) for d in request.destinations])

    record = EventRecord(
        event_id=event_id,
        event_type=request.event_type,
        payload=request.payload,
        created_at=datetime.now(datetime.timezone.utc),
        results=results,
    )
    await store.save(record)

    return DispatchResponse(event_id=event_id, results=list(results))

File: internal/dispatcher/dispatcher.go
package dispatcher

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"math"
	"net/http"
	"sync"
	"time"
	"webhook-dispatcher/internal/formatters"
	"webhook-dispatcher/internal/models"
	"webhook-dispatcher/internal/signing"
	"webhook-dispatcher/internal/store"
)

// Compute delay before attempt
func Backoff(n int, base, max time.Duration) time.Duration {
	delay := time.Duration(float64(base) * math.Pow(2, float64(n)))
	if delay > max {
		return max
	}
	return delay
}

func ptr[T any](v T) *T { return &v }


// Makes one HTTP POST attempt. Returns (httpStatus, errorMsg)
func deliverOnce(
	ctx context.Context,
	client *http.Client,
	url string,
	body map[string]any,
	secret *string,
	timeout time.Duration,
) (int, error) {
	raw, err := json.Marshal(body)
	if err != nil {
		return 0, fmt.Errorf("Marshaling body: %w", err)
	}

	ctx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(raw))
	if err != nil {
		return 0, fmt.Errorf("Building request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	if secret != nil && *secret != "" {
		req.Header.Set("X-Webhook-Signature", signing.Sign(raw, *secret))
	}

	
	resp, err := client.Do(req)
	if err != nil {
		return 0, err
	}
	defer resp.Body.Close()
	io.Copy(io.Discard, resp.Body)
	
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return resp.StatusCode, fmt.Errorf("Non-2xx response: %d", resp.StatusCode)
	}
	return resp.StatusCode, nil
}

// Runs the retry loop for a single destination. Runs inside its own goroutine (launched by Dispatch fn)
func deliverWithRetries(
	ctx context.Context,
	client *http.Client,
	dest models.Destination,
	body map[string]any,
	globalSecret *string,
	maxAttempts int,
	backoffBase, backoffMax, deliveryTimeout time.Duration,
) models.DestinationResult {
	url := dest.URL
	effectiveSecret := dest.Secret
	if effectiveSecret == nil {
		effectiveSecret = globalSecret
	}

	var attempts []models.DeliveryAttempt

	for n := 1; n <= maxAttempts; n++ {
		httpStatus, err := deliverOnce(ctx, client, url, body, effectiveSecret, deliveryTimeout)

		attempt := models.DeliveryAttempt{
			AttemptNumber: n,
			Timestamp:     time.Now().UTC(),
		}
		if err == nil {
			attempt.Status = models.StatusSuccess
			attempt.HTTPStatus = ptr(httpStatus)
		} else {
			attempt.Status = models.StatusFailure
			if httpStatus != 0 {
				attempt.HTTPStatus = ptr(httpStatus)
			}
			attempt.Error = ptr(err.Error())
			log.Printf("Delivery attempt %d to %s failed: %v", n, url, err)
		}
		attempts = append(attempts, attempt)

		if err == nil {
			log.Printf("Delivered to %s on attempt %d", url, n)
			break
		}

		if n < maxAttempts {
			delay := Backoff(n-1, backoffBase, backoffMax)
			log.Printf("Retrying %s in %s", url, delay)
			time.Sleep(delay)
		}
	}

	finalStatus := attempts[len(attempts)-1].Status
	return models.DestinationResult{
		DestinationType: dest.Type,
		URL:             url,
		Status:          finalStatus,
		Attempts:        attempts,
	}
}

// Fans out delivery attempts to all destinations, using one goroutine per destination
func Dispatch(
	ctx context.Context,
	req models.DispatchRequest,
	eventID string,
	client *http.Client,
	s *store.EventStore,
	maxAttempts int,
	backoffBase, backoffMax, deliveryTimeout time.Duration,
) models.DispatchResponse {
	results := make([]models.DestinationResult, len(req.Destinations))
	var wg sync.WaitGroup

	for i, dest := range req.Destinations {
		wg.Add(1)
		go func(i int, dest models.Destination) {
			defer wg.Done()
			body := formatters.Format(dest.Type, req.Payload, req.EventType)
			results[i] = deliverWithRetries(
				ctx, client, dest, body, req.Secret,
				maxAttempts, backoffBase, backoffMax, deliveryTimeout,
			)
		}(i, dest)
	}

	wg.Wait()

	s.Save(models.EventRecord{
		EventID: eventID,
		EventType: req.EventType,
		Payload: req.Payload,
		CreatedAt: time.Now().UTC(),
		Results: results,
	})

	return models.DispatchResponse{EventID: eventID, Results: results}
}